### PR TITLE
Actualtest create card

### PR DESCRIPTION
**Pull request by ChriJ**

### CODE REVIEW

File: Client/Card/index.ts [hunk 1]
```diff
@@ -2,26 +2,29 @@ import * as model from "../../model"
 import { Connection } from "../Connection"
 import { Resource } from "../Resource"
 
-export class Card extends Resource<model.CardResponseV2, model.CreateCardRequest> {
-	constructor(connection: Connection, folder: string, backend: model.CardResponseV2) {
+export class Card extends Resource<model.CardResponseV2 | model.CardResponse, model.CreateCardRequest> {
```
Comment (lucy-meachem) [inline Client/Card/index.ts]: is this actually correct? the response classes are pretty different

File: Client/Cards/create/factory.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+			cardType: result.cardType.cardTypeId,
+			cardNumber: expect.stringMatching(/\d{16}/),
+			cvv: expect.stringMatching(/\d{3}/),
+			expiryDate: expect.stringMatching(/\d{4}-\d{2}/),
+			nameOnCard: expect.any(String),
+			balance: result.balance,
+			issueDate: expect.stringMatching(/\d{4}-\d{2}-\d{2}/),
+			providerCardId: expect.any(String),
+			providerCode: result.providerCode,
+			usage: result.providerCode == "modulr" ? "SINGLE_USE_ALLOW_TEST_AUTH" : "SINGLE_USE",
```
Comment (lucy-meachem) [inline Client/Cards/create/factory.ts:28]: why?

File: Client/Cards/create/factory.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+			createdBy: expect.any(String),
```
Comment (lucy-meachem) [inline Client/Cards/create/factory.ts]: any?

File: Client/Cards/create/factory.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	return [
+		result,
+		{
+			cardType: result.cardType.cardTypeId,
+			cardNumber: expect.stringMatching(/\d{16}/),
+			cvv: expect.stringMatching(/\d{3}/),
+			expiryDate: expect.stringMatching(/\d{4}-\d{2}/),
+			nameOnCard: expect.any(String),
+			balance: result.balance,
+			issueDate: expect.stringMatching(/\d{4}-\d{2}-\d{2}/),
```
Comment (lucy-meachem) [inline Client/Cards/create/factory.ts]: todays date

File: Client/Cards/create/factory.ts (hunk 2 above)
Comment (lucy-meachem) [reply Client/Cards/create/factory.ts:28]: also isnt this a defined enum

File: Client/Cards/create/factory.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import * as model from "../../../model"
+
+export function factory(card: Partial<model.CreateCardRequest>): [model.CreateCardRequest, any, any] {
```
Comment (lucy-meachem) [inline Client/Cards/create/factory.ts]: is any correct?

File: Client/Cards/create/fake.spec.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	)
+	for (const currency of ["EUR", "GBP", "SEK", "JPY", "BRL"])
+		for (const cardType of ["VISA_CREDIT", "MASTERCARD"])
+			it(`card ${currency} ${cardType}`, async () => {
+				const [request, expectedV2, expectedLegacy] = factory({
+					cardType: {
+						cardTypeId: cardType,
+					},
+					currency: currency,
+					providerAccountId: process.env[`accountFake${currency.charAt(0)}${currency.toLowerCase().slice(1)}`],
```
Comment (lucy-meachem) [inline Client/Cards/create/fake.spec.ts:25]: why this?

File: Client/Cards/create/modulr.spec.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+					currency: currency,
+					providerAccountId: process.env[`accountModulr${currency.charAt(0)}${currency.toLowerCase().slice(1)}`],
+					providerCode: "modulr",
+				})
+
+				const cardV2 = await new Promise(resolve => {
+					return setTimeout(() => {
+						resolve(client?.cards.create(request))
+					}, 5000)
+				})
```
Comment (lucy-meachem) [inline Client/Cards/create/modulr.spec.ts]: why does this one wait 5 seconds and nothing else does?

File: Client/Cards/create/fake.spec.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import * as dotenv from "dotenv"
+import * as pax2pay from "../../../index"
+import { factory } from "./factory"
+
+dotenv.config()
+jest.setTimeout(200000)
+
+describe("pax2pay.cards.create fake", () => {
```
Comment (lucy-meachem) [inline Client/Cards/create/fake.spec.ts:8]: these tests are all copied and pasted with almost identical code

File: Client/Cards/create/ixaris.spec.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+dotenv.config()
+jest.setTimeout(200000)
+
+describe("pax2pay.cards.create ixaris", () => {
+	const client = pax2pay.Client.create(process.env.url)
+	for (const currency of ["EUR", "GBP", "USD"])
+		for (const cardType of ["VISA_DEBIT", "VISA_CREDIT", "MASTERCARD"])
+			it(`card ${currency} ${cardType}`, async () => {
+				//authing inside test to reach trackingId for unique friendly names, maybe there's another way
```
Comment (lucy-meachem) [inline Client/Cards/create/ixaris.spec.ts]: i dont understand why you're doing this? the combination of "cardType + currency," is going to be unique for each one regardless.

File: Client/Card/index.ts [hunk 10]
```diff
@@ -2,26 +2,29 @@ import * as model from "../../model"
...
+	constructor(connection: Connection, folder: string, backend: model.CardResponseV2 | model.CardResponse) {
 		super(connection, folder, backend)
 	}
 	amend(request: model.AmendCardRequest) {
-		return this.connection.post<model.CardResponseV2>(`${this.folder}/amend`, request)
+		return this.connection.post<model.CardResponseV2 | model.CardResponse>(`${this.folder}/amend`, request)
```
Comment (lucy-meachem) [inline Client/Card/index.ts]: these classes were all made incorrectly, none of the endpoints other than /v2/cards/virtual makes a v2 response

File: Client/Cards/create/modulr.spec.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+describe("pax2pay.cards.create modulr", () => {
+	const client = pax2pay.Client.create(process.env.url)
+	beforeAll(
+		async () =>
+			await client?.auth.login({
+				username: process.env.username ?? "user",
+				password: process.env.password ?? "password",
+			})
+	)
+	for (const currency of ["EUR", "GBP"])
```
Comment (lucy-meachem) [inline Client/Cards/create/modulr.spec.ts]: modulr also has pln and usd

**Review by lucy-meachem — CHANGES_REQUESTED** [review]

File: Client/Cards/create/modulr.spec.ts (hunk 11 above)
Comment (ChriJ) [reply Client/Cards/create/modulr.spec.ts]: can't find those accounts from searching

File: Client/Cards/create/modulr.spec.ts (hunk 7 above)
Comment (ChriJ) [reply Client/Cards/create/modulr.spec.ts]: modulr gives forbidden error otherwise

**Comment by lucy-meachem** [comment]
ive somehow lost the review thing
but yeah, merge it

**Pull request merged**